nflow-tests/src/main/java/io/nflow/tests/demo/workflow/BulkWorkflow.java

```diff
@@ -0,0 +1,149 @@
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.EnumSet;
+import java.util.List;
+
+import javax.inject.Inject;
+
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.springframework.stereotype.Component;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.nflow.engine.service.WorkflowInstanceService;
+import io.nflow.engine.workflow.definition.NextAction;
+import io.nflow.engine.workflow.definition.StateExecution;
+import io.nflow.engine.workflow.definition.StateVar;
+import io.nflow.engine.workflow.definition.WorkflowDefinition;
+import io.nflow.engine.workflow.definition.WorkflowStateType;
+import io.nflow.engine.workflow.instance.WorkflowInstance;
+import io.nflow.engine.workflow.instance.WorkflowInstance.WorkflowInstanceStatus;
+
+/**
+ * Bulk child workflow executor that does not overflow the system.
+ */
+@Component
+public class BulkWorkflow extends WorkflowDefinition<BulkWorkflow.State> {
+
+  public static final String BULK_WORKFLOW_TYPE = "bulk";
+  public static final String VAR_CHILD_DATA = "childData";
+  public static final String VAR_CONCURRENCY = "concurrency";
+
+  private static final EnumSet<WorkflowInstanceStatus> RUNNING_STATES = complementOf(EnumSet.of(finished, created));
+  private static final Logger logger = getLogger(BulkWorkflow.class);
+
+  @Inject
+  WorkflowInstanceService instanceService;
+
+  public enum State implements io.nflow.engine.workflow.definition.WorkflowState {
+    splitWork(start), waitForChildrenToFinish(normal), done(end), error(manual);
+
+    private WorkflowStateType type;
+
+    State(WorkflowStateType type) {
+      this.type = type;
+    }
+
+    @Override
+    public WorkflowStateType getType() {
+      return type;
+    }
+
+    @Override
+    public String getDescription() {
+      return name();
+    }
+  }
+
+  protected BulkWorkflow(String type) {
+    super(type, splitWork, error);
+    setDescription("Executes child workflows in bulk but gracefully without effecting non-bulk tasks.");
+    permit(splitWork, waitForChildrenToFinish);
+    permit(waitForChildrenToFinish, done);
+  }
+
+  public BulkWorkflow() {
+    this(BULK_WORKFLOW_TYPE);
+  }
+
+  public NextAction splitWork(StateExecution execution, @StateVar(value = VAR_CHILD_DATA, readOnly = true) JsonNode data) {
+    boolean childrenFound = splitWorkImpl(execution, data);
+    if (childrenFound) {
+      return moveToState(waitForChildrenToFinish, "Running");
+    }
+    return retryAfter(waitForChildrenUntil(), "Waiting for child workflows");
+  }
+
+  protected boolean splitWorkImpl(StateExecution execution, @SuppressWarnings("unused") JsonNode data) {
+    if (execution.getAllChildWorkflows().isEmpty()) {
+      throw new RuntimeException(
+          "No child workflows found - either add them before starting the parent or implement splitWorkflowImpl");
+    }
+    return true;
+  }
+
+  protected DateTime waitForChildrenUntil() {
+    return now().plusMinutes(1);
```

Comment:
I would like this to be much longer, preferably an 15min to 60min. When you have 1000 bulk workflows the constant load of just waking without child wakeup will just cause extra load (many wakeups per second). This should just be a workaround for lost wakeup signal. The 1 minute is usually too slow for actual progress without wakeups.

Reply:
Increased to 1 hour